sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashSemiJoin.scala

```diff
@@ -32,34 +32,45 @@ trait HashSemiJoin {
 
   override def output: Seq[Attribute] = left.output
 
-  @transient protected lazy val rightKeyGenerator: Projection =
-    newProjection(rightKeys, right.output)
+  protected[this] def supportUnsafe: Boolean = {
+    (self.codegenEnabled && UnsafeProjection.canSupport(leftKeys)
+      && UnsafeProjection.canSupport(rightKeys)
+      && UnsafeProjection.canSupport(left.schema))
+  }
+
+  override def outputsUnsafeRows: Boolean = right.outputsUnsafeRows
+  override def canProcessUnsafeRows: Boolean = supportUnsafe
```

Comment:
Actually, one question: in the cases where we support UnsafeRows do we always want to force this to use the Unsafe execution path even if the inputs are both safe rows?  If so, I think that we may want to override `canProcessSafeRows` and set it to `!supportUnsafe` in order to force the unsafe path to be taken whenever it can be taken.

As the code exists now, I think that it will choose the Unsafe path as long as either `left` or `right` already happens to output Unsafe before converters have been applied.